### ISSUE TITLE
Problems building Win Client

### ISSUE THREAD

**Issue by josvo** [grke/burp#645]
Hello

I tried building the Windows client for burp, however I ran in a couple of errors, so I wanted to let you know: I'm using Debian Stretch.
When executing ./build-script in burp-cross-tools, the first error is with binutils:

```
source/binutils-2.24/gas/write.c:2295:17: error: left shift of negative value [-Werror=shift-negative-value]
mask = ~((~0) << alignment);
```

=> I could fix this by updating binutils to the version 2.29.1

Then there is an error with gcc:
```
In file included from ../../source/gcc-4.8.0/gcc/cp/except.c:1005:0:
cfns.gperf: In function ‘const char* libc_name_p(const char*, unsigned int)’:
cfns.gperf:101:1: error: ‘const char* libc_name_p(const char*, unsigned int)’ redeclared inline with ‘gnu_inline’ attribute
```
So i tried to update gcc to 6.3 (because debian uses 6.3 as well). I get a bit further, but then there are lots of these errors:
```
In file included from ../../source/mingw-w64-v2.0.9/mingw-w64-crt/intrincs/ilockcxchptr.c:1:0:
../burp-cross-tools/mingw-w64-i686/i686-w64-mingw32/include/intrin.h:894:5: Error: In Konflikt stehende Typen für »_mm_moveldup_ps«
     __MACHINEX86X(__m128 _mm_moveldup_ps(__m128))
```
Any idea? I could not find a solution so far. 
So thats where I'm stuck at the moment. 

What about creating an own branch for the windows-cross-building? I could then update the documentation - as soon as I manage to solve these errors...

Best regards

**Comment by grke** [comment]
Hello,

Sorry, I don't have any ideas why your cross compilation isn't working.
It is very hard to find working combinations of gcc/mingw/binutils, etc.

If I had a separate repo for the build scripts, I think I would need to make the build scripts download the packages from the internet and then check checksums, because it would be a bit ridiculous to have 100MB+ of other people's packages included.

**Comment by josvo** [comment]
I had some success building burp for windows: I could build it with the suggested packages and the documentation provided by you under Centos7. However, I had to patch some vss files manually, because the patch failed. But at least it worked. 

Regarding cross compilation, i think we have to differ 3 things:
1) Building the mingw build environment
2) Building the burp dependent packages
3) Building burp itself. 

I could not manage the first one in debian, with the provided build packages, but the same worked for Centso7. So that's ok. 

It doesn't make sense to provide other peoples packages - that's very true. So i searched a bit and tried some things. 
For 1) there is a build script here:
https://github.com/Zeranoe/mingw-w64-build/blob/master/mingw-w64-build
I tried it in debian works great. So with this script, there would be an actual mingw build environment compared to the one used now. 
It was also possible to build 2) with by slightly modifing the path. 
However, I could not finish building the windows.exe up to now. 

What do you think generally about such an approach? We would need some scripts, the patches and also some readme files.

**Comment by grke** [comment]
Hello,

I don't understand what you mean by "there would be an actual mingw build environment compared to the one used now". As far as I can see, that build script does the same job as mine, except that it downloads source and creates a more up-to-date environment (which is great). I think it should use checksums though, to see that what it has downloaded matches what it should download.

Generally, I think the approach is good.
I should just create a new repo with the required things, not including the source packages, and make the scripts download and check checksums.

**Comment by josvo** [comment]
Sorry for not beeing clear. I just wanted to say that I could not build mingw due to errors in Debian 9 with the existing script for burp. But it worked with the updated packages (the script linked above).
If you want I can give it a try:
- take your existing script and change it
- download the packages
- check checksums

**Comment by grke** [comment]
That's OK, I will try to do it this weekend and make the repository for it. Thanks.

**Comment by grke** [comment]
So, this weekend, I changed the existing scripts to be able to work from within the same directory, and to download the sources and check checksums. I then added them to github.

Here they are:
https://github.com/grke/burp-cross-tools

These are still using the old sources, so they won't work on your debian 9 system.
However, maybe it is now easier for you to update them?

**Comment by josvo** [comment]
Thanks for the repository. I really simplifies things. 
I made some progress, so this is what I found out so far:
I downloaded your script and dependencies and tried to build the Windows binaries. 

## Building mingw

The first error is this:

Error
../../source/binutils-2.24/gas/read.c:5022:12: error: left shift of negative value [-Werror=shift-negative-value]
  val |= ~0 << loaded;

I used this patch to avoid the error
https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=commit;h=29798047afaa22335a504280a5bbdbafe2b1b63d

However, the next error was:
../bfd/bfd.h:304:75: error: right-hand operand of comma expression has no effect [-Werror=unused-value]
There is another patch
https://sourceware.org/ml/binutils/2014-01/msg00333.html

But I then realised that the problem is that Debian 9 is using GCC6 and **binutils 2.24 can't be built with GCC > 4.8.** These errors have been fixed with recent binutils versions, so I switched to binutils 2.26.

Then there was an error with building GCC 4.7.4 but this can be fixed with a patch mentioned here:
https://gcc.gnu.org/ml/gcc-patches/2015-08/msg00375.html
The patch I used is attached here: [gcc.txt](https://github.com/grke/burp/files/1527832/gcc.txt)
Obviously here is the same problem as with binutils: **GCC versions > 4.8 are not able to build GCC 4.7.4.** (without patching it)

With these two changes, the build environment can be created. 

## Building the dependencies
This works quite straight forward, however the vss.patches do not apply cleanly so I had to modifiy them. I had to
delete the "pragma" in vss.h
delete the "typedef" in vswriter.h

## Building the binaries
This finally worked and I have working binaries now!

So long story short: 
The problem is not within Debian but with the used GCC6 which won't build the provided binutils and GCC version. As more and more systems use more actual compilers, this might become a problem. 
That's why I suggest to use a more recent GCC version and also depks and especially mingw version. I now this might not be easy, but I can try it and report back.

**Comment by josvo** [comment]
Sorry, did not mean to close it, as there are some patches which should eventually be applied. I would like to leave this up to you.
But building now works, which is a success.

**Comment by grke** [comment]
That's OK.
I was expecting that an outcome of this was for the burp windows install build tools to be upgraded to newer/latest versions of packages, so upgrading to newer versions is fine. It just means that I need to upgrade my build machine too.

**Comment by Zouloc** [comment]
Hi,
Sorry for posting here (don't know if it is the right place to post)...
First, I want to thank both of you: All your information are useful to build mingw, by combining burp-cross-tools github repo and @josvo 's information it builds fine! 
Secondly, I think it might be useful to update burp-cross-tools github in order to add @josvo 's solution (switch binutils to 2.26 and apply the patch to GCC).
Finally, my problem... I'm missing the depkgs/source/vss.tar.gz. It's not present in the github repo and the building script doesn't download it.
So here is my question, How do I get it?
Thank for helping

Regards

**Comment by josvo** [comment]
There was a file called README.VSS in the depkgs. You can find a version here: https://burp.grke.org/downloads/windows-build-tools/ (Download a version of burp-depkgs and unpack it, then you will find it). 
The file explains how to get the VSS files. Due to licensing problems you have to download them yourself and they probably can't be set in a repository.
@grke might be a good idea to put the README.VSS back to the repo, what do you think?

**Comment by Zouloc** [comment]
README.VSS is in the repo... I just missed it!
Thanks for helping and so sorry for the disturbance...

**Comment by grke** [comment]
No problem, thanks for the kind words. :)

It's a reminder that I NTH

**Comment by grke** [comment]
Haha, my keyboard actually broke as I was typing!

I meant to write:
"It's a reminder that I need to upgrade my build process with a new build server."

**Comment by Zouloc** [comment]
Urgl! You hit your keyboard too hard!
Damn... Did I miss something else (Am I blind?!?)... I'm getting another error, It looks like I'm missing other dependencies...
«check_log.c:27:27: fatal error: subunit/child.h: No such file or directory»
Sorry (again) for the disturbance :(

**Comment by Zouloc** [comment]
Ok! I find the solution... It seems I had a conflict between build environment and installed packages...
«apt remove libsubunit-dev» solved the problem!

**Comment by grke** [comment]
I'm currently working on this.
I'm going to try to make it work for debian stretch running gcc 6, compiling gcc 7.3.0 and mingw-w64 5.0.3. The posts of josvo above are super useful for doing that.

**Comment by josvo** [comment]
I'm glad I could help! 
@grke if you want me to test something or I can help anyhow, just tell me! I remember trying your suggested environment once but I did not succeed. So if you are stuck somewhere, I could give it another try, just post your findings. 
@Zouloc if you found out something while you were building the binaries that is worth mentioning here, please do not hesitate to post it. Also, perhaps you would like to help testing, as soon there is a working environment for newer versions of gcc and mingw?

**Comment by Zouloc** [comment]
Hi,
Sorry for the delay, I'm working on multiple projects.
With all the informations above, I successfully built depkgs and mingw, but I don't know what to do next with those binaries (It's my 1st win build) ^^'
I would be glad to help testing :)

**Comment by grke** [comment]
I too have built cross-tools and depkgs with the new versions.
I found that I needed to make a few changes to burp itself in order to get a successful build of the installers.
Those changes are now in burp git master, except for one.
That last change will break the current Windows installer builds, so I am not committing it just yet.
I also haven't tested the newly built installers at all.

What is good about my new cross-tools/depkg build system is that I have done it using Docker.
That means that you should be able to very easily reproduce what I have done, using a minimal number of commands.

Next step is to hook up the docker image to my automatic build system and have it take over from the old system.

**Comment by grke** [comment]
I will push the burp-cross-tools changes to the origin/debian branch shortly.

**Comment by grke** [comment]
Done. This is the uncommitted patch to burp that lets it build the installers:

```
diff --git a/src/win32/Makefile.installer b/src/win32/Makefile.installer
index 56faa732..53cd2aca 100644
--- a/src/win32/Makefile.installer
+++ b/src/win32/Makefile.installer
@@ -7,9 +7,11 @@ include ../Makefile.inc
 ifeq ($(WIN64),yes)
   SSL_DLL := libssl-1_1-x64.dll
   CRYPTO_DLL := libcrypto-1_1-x64.dll
+  LIBGCC_DLL := libgcc_s_seh-1.dll
 else
   SSL_DLL := libssl-1_1.dll
   CRYPTO_DLL := libcrypto-1_1.dll
+  LIBGCC_DLL := libgcc_s_sjlj-1.dll
 endif
 
 DEFINES := \
@@ -22,6 +24,7 @@ DEFINES := \
        -DBITS=$(BITS) \
        -DSSL_DLL=$(SSL_DLL) \
        -DCRYPTO_DLL=$(CRYPTO_DLL) \
+       -DLIBGCC_DLL=$(LIBGCC_DLL) \
 
 INSTALL_EXE := $(BUILDDIR)/release$(BITS)/$(PACKAGE_TARNAME)-win$(BITS)-installer-$(PACKAGE_VERSION).exe
 
@@ -42,7 +45,7 @@ DEPKGS_BINARIES := \
        openssl.exe
 
 MINGW_LIBS := \
-       libgcc_s_sjlj-1.dll
+       $(LIBGCC_DLL)
 
 DOC_FILES :=
 
diff --git a/src/win32/installer/winburp.nsi b/src/win32/installer/winburp.nsi
index 8708399c..8aa6c0fe 100644
--- a/src/win32/installer/winburp.nsi
+++ b/src/win32/installer/winburp.nsi
@@ -307,7 +307,7 @@ Function InstallCommonFiles
                File "${SRC_DIR}\libpcre-1.dll"
                File "${SRC_DIR}\libyajl.dll"
                File "${SRC_DIR}\zlib1.dll"
-               File "${SRC_DIR}\libgcc_s_sjlj-1.dll"
+               File "${SRC_DIR}\${LIBGCC_DLL}"
                File "${SRC_DIR}\compat.dll"
                File "${SRC_DIR}\libcheck-0.dll"
                File "${SRC_DIR}\openssl.exe"
```

**Comment by grke** [comment]
Hello,

All the changes are now in burp-cross-tools master, and burp master, and my automatic build system uses a docker image created from burp-cross-tools.

It is based on debian stretch, and uses these packages:
binutils-2.29.1
gcc-7.3.0
isl-0.18
mingw-w64-v5.0.3
mpc-1.1.0
mpfr-4.0.1

As everything is up-to-date now, and simple for everybody to replicate, I am closing this issue.
Thanks for all your help, it took a while but I think it was worth it.
